sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/FilterEstimation.scala

```diff
@@ -0,0 +1,569 @@
+            val newValue = Some(literal.value)
+            aColStat.copy(distinctCount = 1, min = newValue,
+              max = newValue, nullCount = 0)
+          case DateType =>
+            val dateValue = Some(literal.value)
+            aColStat.copy(distinctCount = 1, min = dateValue,
+              max = dateValue, nullCount = 0)
+          case TimestampType =>
+            val tsValue = Some(literal.value)
+            aColStat.copy(distinctCount = 1, min = tsValue,
+              max = tsValue, nullCount = 0)
+          case _ => aColStat.copy(distinctCount = 1, nullCount = 0)
+        }
+        mutableColStats += (attrRef.exprId -> newStats)
+      }
+
+      Some(1.0 / ndv.toDouble)
+    } else {
+      Some(0.0)
+    }
+
+  }
+
+  /**
+   * Returns a percentage of rows meeting "IN" operator expression.
+   * This method evaluates the equality predicate for all data types.
+   *
+   * @param attrRef an AttributeReference (or a column)
+   * @param hSet a set of literal values
+   * @param update a boolean flag to specify if we need to update ColumnStat of a given column
+   *               for subsequent conditions
+   * @return a doube value to show the percentage of rows meeting a given condition
+   *         It returns None if no statistics exists for a given column.
+   */
+
+  def evaluateInSet(
+      attrRef: AttributeReference,
+      hSet: Set[Any],
+      update: Boolean)
+    : Option[Double] = {
+    if (!mutableColStats.contains(attrRef.exprId)) {
+      logDebug("[CBO] No statistics for " + attrRef)
+      return None
+    }
+
+    val aColStat = mutableColStats(attrRef.exprId)
+    val ndv = aColStat.distinctCount
+    val aType = attrRef.dataType
+
+    // use [min, max] to filter the original hSet
+    val validQuerySet = aType match {
+      case _: NumericType | DateType | TimestampType =>
+        val statsRange =
+          Range(aColStat.min, aColStat.max, aType).asInstanceOf[NumericRange]
+        val hSetMap = hSet.map(e =>
+            if (e.isInstanceOf[String]) {
+              val utf8String = UTF8String.fromString(e.asInstanceOf[String])
+              numericLiteralToBigDecimal(aType, utf8String, StringType)
+            } else {
+              numericLiteralToBigDecimal(aType, e, aType)
+            })
+        hSetMap.filter(e => e >= statsRange.min && e <= statsRange.max)
+
+      // We assume the whole set since there is no min/max information for String/Binary type
+      case StringType | BinaryType => hSet
+    }
+    if (validQuerySet.isEmpty) {
+      return Some(0.0)
+    }
+
+    // newNdv should be no greater than the old ndv.  For example, column has only 2 values
+    // 1 and 6. The predicate column IN (1, 2, 3, 4, 5). validQuerySet.size is 5.
+    val newNdv = math.min(validQuerySet.size.toLong, ndv.longValue())
+    val(newMax, newMin) = aType match {
+      case _: NumericType =>
+        val tmpSet: Set[Double] = validQuerySet.map(e => e.toString.toDouble)

```

Comment:
This breaks the assumption of `ColumnStats`: we should store Long for integral type, Double for float/double type, Decimal for decimal type.

Maybe we should have a method `BigDecimalToNumeric`